src/engraving/libmscore/shape.cpp

```diff
@@ -102,6 +102,24 @@ bool Shape::sameVoiceExceptions(const EngravingItem* item1, const EngravingItem*
     return false;
 }
 
+bool Shape::graceToMainExceptions(const EngravingItem* item1, const EngravingItem* item2) const
+{
+    // Handles exceptions in the cases of main-to-grace and grace-to-main note spacing
```

Comment:
I'd suggest you to rewrite it on the following manner:

if (!item1 || !item1->isNote()) {
    return false;
}

....

filter out the very common cases first and then go down to details

Reply:
you're free to change it in the next PRs, tho